File: src/runtime/boxes/capture/bb_capture.java
package bb;

/**
 * bb_capture.java — CAPTURE: wrap child; $ writes on every γ; . buffers for Phase-5 commit
 * Port of bb_capture.c / bb_capture.s
 *
 *   CAP_α:       cr = child(α);  if empty → CAP_ω;  else → CAP_γ_core;
 *   CAP_β:       cr = child(β);  if empty → CAP_ω;  else → CAP_γ_core;
 *   CAP_γ_core:  if (immediate) NV_SET(varname, cr);
 *                else           pending=cr; has_pending=true;
 *                               return cr;
 *   CAP_ω:       has_pending=false; return spec_empty;
 *
 * The executor calls commitPending() after overall match success to flush
 * deferred (.) captures — mirroring match.clj commit-pending! and stmt_exec.c.
 */
public class bb_capture extends bb_box {

    /** Callback interface for variable assignment (NV_SET_fn equivalent) */
    public interface VarSetter {
        void set(String varname, String value);
    }

    private final bb_box     child;
    private final String    varname;
    private final boolean   immediate;   /* true = $var; false = .var */
    private final VarSetter setter;

    /* deferred capture state (.var path) */
    private Spec    pending    = null;
    private boolean hasPending = false;

    public bb_capture(MatchState ms, bb_box child, String varname,
                     boolean immediate, VarSetter setter) {
        super(ms);
        this.child     = child;
        this.varname   = varname;
        this.immediate = immediate;
        this.setter    = setter;
    }

    @Override public Spec α() { return runChild(child.α()); }
    @Override public Spec β()  { return runChild(child.β());  }

    private Spec runChild(Spec cr) {
        if (cr == null) {
            // CAP_ω
            hasPending = false;
            return null;
        }
        // CAP_γ_core
        if (varname != null && !varname.isEmpty()) {
            String matched = ms.sigma.substring(cr.start, cr.start + cr.len);
            if (immediate) {
                setter.set(varname, matched);                                  // $ — write now
            } else {
                pending    = cr;                                               // . — buffer
                hasPending = true;
            }
        }
        return cr;
    }

    /**
     * Called by executor after overall match succeeds to commit deferred captures.
     * Mirrors match.clj commit-pending! and stmt_exec.c Phase-5 flush.
     */
    public void commitPending() {
        if (hasPending && pending != null && varname != null && !varname.isEmpty()) {
            setter.set(varname, ms.sigma.substring(pending.start, pending.start + pending.len));
            hasPending = false;
            pending    = null;
        }
    }
}

File: src/runtime/boxes/capture/bb_capture.cs
// bb_capture.cs — CAPTURE: $ writes on every γ; . buffers for Phase-5 commit
// Mirrors src/runtime/boxes/bb_capture.c
//
// Wraps a child box.  On every γ (α or β):
//   immediate=true  ($var): write matched substring to variable NOW
//   immediate=false (.var): buffer pending capture; Phase 5 commits on :S
//
// On ω: clear pending.
//
// The VariableStore delegate matches snobol4dotnet's IdentifierTable write path.

namespace Snobol4.Runtime.Boxes;

public sealed class bb_capture : IByrdBox
{
    private readonly IByrdBox _child;
    private readonly string   _varname;
    private readonly bool     _immediate;

    // Injected by the executor — writes varname=value into IdentifierTable
    public  Action<string, string>? SetVar    { get; set; }
    // Pending capture set on every γ; flushed by Phase 5 on :S
    public  string?                 Pending   { get; private set; }
    public  bool                    HasPending { get; private set; }

    public bb_capture(IByrdBox child, string varname, bool immediate)
    {
        _child     = child;
        _varname   = varname ?? "";
        _immediate = immediate;
    }

    public Spec Alpha(MatchState ms) => TryChild(ms, alpha: true);
    public Spec Beta(MatchState ms)  => TryChild(ms, alpha: false);

    private Spec TryChild(MatchState ms, bool alpha)
    {
        var cr = alpha ? _child.Alpha(ms) : _child.Beta(ms);
        if (cr.IsFail)
        {
            HasPending = false;
            return Spec.Fail;                                       // CAP_ω
        }

        // γ path — extract matched substring
        string matched = ms.Subject.Substring(cr.Start, cr.Length);

        if (_immediate && !string.IsNullOrEmpty(_varname))
        {
            SetVar?.Invoke(_varname, matched);                      // $ immediate write
        }
        else if (!string.IsNullOrEmpty(_varname))
        {
            Pending    = matched;                                   // . deferred write
            HasPending = true;
        }

        return cr;                                                  // CAP_γ
    }

    // Called by Phase 5 on :S to commit pending conditional captures
    public void CommitPending()
    {
        if (HasPending && !string.IsNullOrEmpty(_varname))
            SetVar?.Invoke(_varname, Pending ?? "");
        HasPending = false;
    }
}
